src/App/Fossa/Analyze/Upload.hs

```diff
@@ -100,13 +102,15 @@ uploadSuccessfulAnalysis (BaseDir basedir) metadata jsonOutput revision units li
 
     dieOnMonorepoUpload revision
 
-    uploadResult <- case licenseUnits of
-      Nothing -> uploadAnalysis revision metadata units
-      Just licenses -> do
+    uploadResult <- case (maybeUnits, maybeLicenseUnits) of
+      (Just units, Nothing) -> uploadAnalysis revision metadata units
+      (mUnits, Just licenses) -> do
         org <- getOrganization
         let fullFileUploads = FullFileUploads $ orgRequiresFullFileUploads org
+        let units = maybe [] NE.toList mUnits
         let mergedUnits = mergeSourceAndLicenseUnits units licenses
         runStickyLogger SevInfo $ uploadAnalysisWithFirstPartyLicensesToS3AndCore revision metadata mergedUnits fullFileUploads
+      (Nothing, Nothing) -> fatalText "Attempted to upload with no results" -- This should never happen, as the caller to this function checks for this before calling
```

Comment:
Since this comment imposes a responsibility on the caller I think that it should be repeated in a comment docstring.

Alternatively, I think this type forbids what you're talking about and makes the type checker force the caller to do what's correct if you accept it as an argument to the function:
```haskell
data ScanUnits =
SourceUnitsOnly (NE.NonEmpty SourceUnit)
| LicenseSourceUnitOnly LicenseSourceUnit
| SourceAndLicenseUnits (NE.NonEmpty SourceUnit) LicenseSourceUnit
```